beets/library.py

```diff
@@ -715,6 +715,71 @@ def move(self, copy=False, link=False, basedir=None, with_album=True):
 
     # Templating.
 
+    def _legalize_partial(self, path, fragment, replacements):
+        """ Perform a partial legalization of the path (ie. a single
+        sanitization and truncation). Outputs unicode if fragment is set,
+        otherwise bytes.
+        """
+
+        # Truncate components and remove forbidden characters.
+        path = util.sanitize_path(path, replacements)
+
+        # Encode for the filesystem.
+        if not fragment:
+            path = bytestring_path(path)
+
+        # Preserve extension.
+        _, extension = os.path.splitext(self.path)
+        if fragment:
+            # Outputting Unicode.
+            extension = extension.decode('utf8', 'ignore')
+        path += extension.lower()
+
+        # Truncate too-long components.
+        maxlen = beets.config['max_filename_length'].get(int)
+        if not maxlen:
+            # When zero, try to determine from filesystem.
+            maxlen = util.max_filename_length(self._db.directory)
+        path = util.truncate_path(path, maxlen)
+
+        return path
+
+    def _legalize_path(self, path, fragment):
+        """ Perform several iterations of _legalize_partial, to generate a
+        stable, optimal output path. This is necessary for cases where
+        truncation produces unclean paths (eg. trailing space).
+        """
+
+        # Create a list of path candidates
+        path_candidates = [b'']
+
+        replacements = self._db.replacements
+
+        # Perform an initial pass
+        path = self._legalize_partial(path, fragment, replacements)
+        while path != path_candidates[-1]:
+            # This will keep sanitizing the path until it's stable, or an
+            # infinite loop appears
+            while path not in path_candidates:
```

Comment:
Yeah, I totally agree that we should not prematurely optimize. This is quite unlikely to be a bottleneck. :smiley: I was just proposing the simpler one-shot approach for predictability, not really performance. Since we can't prove that weird, unbounded, acyclic fluctuations won't happen, I thought that a simple limit would make it certain that this doesn't run for an arbitrary (finite) amount of time.

I totally get it if you think this version is simpler or more maintainable, though, so I'll defer to your preference.